controllers/kafkacluster_controller.go

```diff
@@ -147,84 +142,167 @@ func (r *KafkaClusterReconciler) Reconcile(request ctrl.Request) (ctrl.Result, e
+			return requeueWithError(log, "failed to remove topics finalizer from kafkacluster", err)
+		}
+	}
+
+	// If any of the topics still exist, it means their finalizer is still running.
+	// Wait to make sure we have fully cleaned up zookeeper. Also if we delete
+	// our kafkausers before all topics are finished cleaning up, we will lose
+	// our controller certificate.
+	log.Info("Ensuring all topics have finished cleaning up")
+	var childTopics v1alpha1.KafkaTopicList
+	if err = r.Client.List(ctx, &childTopics); err != nil {
 		return requeueWithError(log, "failed to list kafkatopics", err)
 	}
 	for _, topic := range childTopics.Items {
 		if belongsToCluster(topic.Spec.ClusterRef, cluster) {
-			log.Info(fmt.Sprintf("Deleting associated topic %s", topic.Name))
-			if err = r.Client.Delete(context.TODO(), &topic); err != nil {
-				return requeueWithError(log, "failed to delete kafkatopic", err)
-			}
+			log.Info(fmt.Sprintf("Still waiting for topic %s/%s to be deleted", topic.Namespace, topic.Name))
+			return ctrl.Result{
+				Requeue:      true,
+				RequeueAfter: time.Duration(3) * time.Second,
+			}, nil
 		}
 	}
 
-	// Delete all kafkausers associated with us
-	var childUsers banzaicloudv1alpha1.KafkaUserList
-	if err = r.Client.List(context.TODO(), &childUsers); err != nil {
-		return requeueWithError(log, "failed to list kafkausers", err)
-	}
-	for _, user := range childUsers.Items {
-		if belongsToCluster(user.Spec.ClusterRef, cluster) {
-			log.Info(fmt.Sprintf("Deleting associated user %s", user.Name))
-			if err = r.Client.Delete(context.TODO(), &user); err != nil {
-				return requeueWithError(log, "failed to delete kafkauser", err)
+	// If we haven't deleted all kafkausers yet, iterate namespaces and delete all kafkausers
+	// with the matching label.
+	if util.StringSliceContains(cluster.GetFinalizers(), clusterUsersFinalizer) {
+		log.Info(fmt.Sprintf("Sending delete kafkausers request to all namespaces for cluster %s/%s", cluster.Namespace, cluster.Name))
+		for _, ns := range namespaces.Items {
+			if err := r.Client.DeleteAllOf(
+				ctx,
+				&v1alpha1.KafkaUser{},
+				client.InNamespace(ns.Name),
+				client.MatchingLabels{clusterRefLabel: clusterLabelString(cluster)},
+			); err != nil {
+				if client.IgnoreNotFound(err) != nil {
+					return requeueWithError(log, "failed to send delete request for children kafkausers", err)
+				}
+				log.Info(fmt.Sprintf("No matching kafkausers in namespace: %s", ns.Name))
 			}
 		}
+		if cluster, err = r.removeFinalizer(ctx, cluster, clusterUsersFinalizer); err != nil {
+			return requeueWithError(log, "failed to remove users finalizer from kafkacluster", err)
+		}
+	}
+
+	// Do any necessary PKI cleanup - a PKI backend should make sure any
+	// user finalizations are done before it does its final cleanup
+	log.Info("Tearing down any PKI resources for the kafkacluster")
+	if err = pki.GetPKIManager(r.Client, cluster).FinalizePKI(ctx, log); err != nil {
```

Comment:
If SSL isn't specified in KafkaCluster CR this will be a nil pointer dereference when the `GetPKIManager` function will be called.

Reply:
Fixing. I'm trying to decide if I want some sort of check in the `KafkaUser` controller also. Creating a `KafkaUser` doesn't make sense if you are not using SSL right now - but looking closer that would probably also panic if someone tried. 

Might be abother good use-case to expand the admission webhook